src/rustc-driver.md

```diff
@@ -1,46 +1,25 @@
-[`Session`] \(general build context and error messaging) and the [`TyCtxt`]
-\(the "typing context", allowing you to query the type system and other cool
-stuff). The `rustc_driver` crate also provides external users with a method
+using the interface defined in the [`rustc_interface`] crate.
+
+The `rustc_interface` crate provides external users with an (unstable) API
 for running code at particular times during the compilation process, allowing
 third parties to effectively use `rustc`'s internals as a library for
-analysing a crate or emulating the compiler in-process (e.g. the RLS).
-
-For those using `rustc` as a library, the `run_compiler()` function is the main
-entrypoint to the compiler. Its main parameters are a list of command-line
-arguments and a reference to something which implements the `CompilerCalls`
-trait. A `CompilerCalls` creates the overall `CompileController`, letting it
-govern which compiler passes are run and attach callbacks to be fired at the end
-of each phase.
-
-From `rustc_driver`'s perspective, the main phases of the compiler are:
-
-1. *Parse Input:* Initial crate parsing
-2. *Configure and Expand:* Resolve `#[cfg]` attributes, name resolution, and
-   expand macros
-3. *Run Analysis Passes:* Run trait resolution, typechecking, region checking
-   and other miscellaneous analysis passes on the crate
-4. *Translate to LLVM:* Translate to the in-memory form of LLVM IR and turn it
-   into an executable/object files
+analysing a crate or emulating the compiler in-process (e.g. the RLS or rustdoc).
 
-The `CompileController` then gives users the ability to inspect the ongoing
-compilation process
+For those using `rustc` as a library, the `interface::run_compiler()` function is the main
+entrypoint to the compiler. It takes a configuration for the compiler and a closure that
+takes a [`Compiler`]. `run_compiler` creates a `Compiler` from the configuration and passes
+it to the closure. Inside the closure, you can use the `Compiler` to drive queries to compile
+a crate and get the results. This is what the `rustc_driver` does too.
 
-- after parsing
-- after AST expansion
-- after HIR lowering
-- after analysis, and
-- when compilation is done
-
-The `CompileState`'s various `state_after_*()` constructors can be inspected to
-determine what bits of information are available to which callback.
-
-For a more detailed explanation on using `rustc_driver`, check out the
-[stupid-stats] guide by `@nrc` (attached as [Appendix A]).
+You can see what queries are currently available through the rustdocs for [`Compiler`].
+You can see an example of how to use them by looking at the `rustc_driver` implementation,
+specifically the [`rustc_driver::run_compiler` function][rd_rc] (not to be confused with
+`interface::run_compiler`). The `rustc_driver::run_compiler` function takes a bunch of
+command-line args and some other configurations and drives the compilation to completion.
```

Comment:
Maybe mention that `rustc_driver::run_compiler` takes a `Callbacks` type, which essentially replaces `CompilerCalls` + `CompileController`?